### PR TITLE
Update condition key, rewrite `ConditionalCodec` implementation and add tests

### LINKED ISSUE #230: The "Elevators" datapack by Vanilla Tweaks is broken with NeoForge
**Minecraft Version:** 1.20.2

**NeoForge Version:** 20.2.20-beta

NeoForge:

https://github.com/neoforged/NeoForge/assets/68275908/865913a8-6c15-41da-a922-3f73ebc4a14d

Vanilla:

https://github.com/neoforged/NeoForge/assets/68275908/51761ff6-9b4b-45ec-aa76-b525d4eb5839

The datapack is supposed to turn wool into an elevator when an Ender pearl is dropped on it.

https://vanillatweaks.net/picker/datapacks/

### PR DESCRIPTION
**Pull request by Technici4n**
Completely rewrite `ConditionalCodec` implementation, adding a lot of documentation and tests. This fixes a number of edge cases that the previous implementation had. It's basically impossible to list them all - we should reason based on what we expect the codec to do from now on, and add corresponding test cases.

**Conditions now always use the `neoforge:conditions` key. Previously this was `conditions` for advancements, recipes, and dynamic registries. Please update your mods accordingly.**

Any non-map value can have conditions in the following format:
```json5
{
    "neoforge:conditions": [...],
    "neoforge:value": // value that needs a condition check
}
```
Any map-like value can have the conditions inline.

For example this is how to serialize a record with a single `value` field:
```json5
// Works: (no condition)
{
    "value": "test"
}
// Works: (condition, nested object)
{
    "neoforge:conditions": [
        { "type": "neoforge:true" }
    ],
    "neoforge:value": {
        "value": "test"
    }
}
// Works: (condition, inline object)
{
    "neoforge:conditions": [
        { "type": "neoforge:true" }
    ],
    "value": "test"
}
```

See the tests for more examples.

Also updates `DataProvider` to sort `neoforge:conditions` before the other keys.

Fixes #230 as a side effect.

### CODE REVIEW

**Comment by lukebemish** [comment]
Looking through this it looks sensible to me - the chain seems to be to first try to use the approach of directly encoding/decoding to a MapCodec nested in the conditional codec, then use the wrapped codec to encode/decode a map and then grab stuff from that, or, in the case of the decoder wrapper, just fall back to the normal unconditional codec if those don't work. ~~In practice, there are one or two edge cases that this could break in, but I'm unsure if those actually exist in any of the locations this is being used at in patches. Basically, is there anywhere in vanilla where the use of this would result in a CodecMapCodec becoming a non-MapCodec-based codec, or vice versa, in a location that's the target of a delegate? Those are the only places I can imagine this causing issues - I'll try to poke around and see whether any of those are currently using this. At any rate, I'd say this does definitely properly diagnose the issue in question and seems like a sensible solution to it.~~ (I can find places in vanilla where this is likely to break stuff)

**Comment by lukebemish** [comment]
On second inspection... this may have some issues. For instance, this patch: <https://github.com/neoforged/NeoForge/blob/1.20.x/patches/net/minecraft/world/level/storage/loot/LootDataType.java.patch#L31-L37> will in general neatly assume that the codec it's passed encodes map-like stuff in some form, after this PR. Do we have a basis for that assumption?

**Comment by lukebemish** [comment]
Okay yeah this will be an issue. Let me write out an example:
- Initializing `LootDataType.PREDICATE` calls the LootDataType ctor that wraps a conditional codec around it
- The _original_ predicate codec, `LootItemConditions.CODEC`, includes an "inline" codec that can accept lists
- Upon wrapping it with `ConditinalOps#createConditionalCodec` and then calling `.codec()` on it in the ctor of `LootDataType`, it instead produces a codec that can _only_ accept or decode maps, as that's all that the constructed map codec could do.
The proper solution to this is to make a ctor that producs a `Codec<Optional<T>>` (**not a MapCodec**), with the following logic, in pseudocode:
```java
public static <T> Codec<Optional<T>> createConditionalCodec(final Codec<T> ownerCodec, String conditionalsKey) {
    if (ownerCodec instanceof MapCodec.CodecMapCodec) {
        return createConditionalMapCodec(...).codec();
    }
    return new Codec() {
        // on encode, encode the ownerCodec first, and then try to add the conditions to it if it's a map and error otherwise
        // on decode, check if it's a map; if not, decode directly with ownerCodec, and if so, check conditions and then decode
    }
}
```
Or something like this: https://gist.github.com/lukebemish/9ce30f932a02f6ce344a1a833d5bf7dd (note, entirely untested, I'm struggling to set up forgedev), which solves the problem by wrappin MapCodecs in MapCodecs and non-MapCodecCodec Codecs in normal old Codecs.

Note that this sort of issue is present with the current stuff too - this PR fixes one issue but doesn't address the rest of the issues with the system

**Comment by Technici4n** [comment]
Rewrote the whole thing again. Hopefully it should now handle all the edge cases.

File: src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java [hunk 1]
```diff
@@ -193,52 +182,55 @@ private ConditionalDecoder(String conditionalsPropertyKey, Codec<List<ICondition
...
+                return conditionsCodec.decode(ops, conditionsDataCarrier).flatMap(conditionsCarrier -> {
+                    final List<ICondition> conditions = conditionsCarrier.getFirst();
+                    final DataResult<Pair<ICondition.IContext, T>> contextDataResult = contextCodec.decode(ops, ops.emptyMap());
+
+                    return contextDataResult.flatMap(contextCarrier -> {
+                        final ICondition.IContext context = contextCarrier.getFirst();
+
+                        final boolean conditionsMatch = conditions.stream().allMatch(c -> c.test(context));
+                        if (!conditionsMatch)
+                            return DataResult.error(() -> "Conditions did not match", Pair.of(Optional.empty(), input));
```
Comment (lukebemish) [inline src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java]: Why is this an error? Shouldn't the codec return a successful, empty optional when conditions don't match?

File: src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java [hunk 2]
```diff
@@ -193,52 +182,55 @@ private ConditionalDecoder(String conditionalsPropertyKey, Codec<List<ICondition
...
+
+                        DataResult<Pair<A, T>> innerDecodeResult;
+
+                        T valueDataCarrier = inputMap.get(valuePropertyKey);
+                        if (valueDataCarrier != null && inputMap.entries().count() == 2) {
```
Comment (lukebemish) [inline src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java]: This isn't the proper logic - this would fail if, for instance, someone wrapped a `someCodec.fieldOf("value").codec()` in a conditional codec. If there's only a single value field, you have to try decoding with *both* the inner decoding directly of the value field and removing the conditional field and decoding that.

File: src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java [hunk 3]
```diff
@@ -61,123 +68,105 @@ public static <T> Codec<List<T>> decodeListWithElementConditionsAndConsumeIndex(
...
-    public static <T> MapCodec<Optional<WithConditions<T>>> createConditionalCodecWithConditions(final Codec<T> ownerCodec, String conditionalsKey) {
-        return new ConditionalCodec<T>(
-                conditionalsKey,
-                ICondition.LIST_CODEC,
-                retrieveContext().codec(),
-                ownerCodec);
+    /**
+     * Creates a conditional codec. The codec assumes that the inner object (of type {@code T}) serializes as a map.
+     */
+    public static <T> Codec<Optional<WithConditions<T>>> createConditionalCodecWithConditions(final Codec<T> ownerCodec, String conditionalsKey) {
```
Comment (lukebemish) [inline src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java:121]: Probably worth adding a comment somewhere here noting the use of a codec created by this in anywhere that is delegated to by a KeyDispatchCodec may have differing behaviour, on decoding, than the original codec, since we don't want someone accidentally using it that way in a neo patch and breaking vanilla compat again in the future

**Review by lukebemish — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java (hunk 3 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java:121]: Because it's not a `MapCodec` you mean?

File: src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java (hunk 2 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java]: Fixed - now `neoforge:value` is used as the value key.

File: src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java (hunk 3 above)
Comment (lukebemish) [reply src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java:121]: Yeah, cause it's not a MapCodecCodec but could wrap one and KeyDispatchCodec is funky. As long as it's documented it should be fine

Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java:121]: Added a bit of javadoc for that in 33eabb0.

File: src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java (hunk 1 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java]: Changed to successful with an empty Optional.

**Comment by Technici4n** [comment]
Note: this needs #250 to be merged first so that we can update it with the new condition key.

EDIT: Done!

File: tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+            JsonElement json = read("""
+                    {
+                        "neoforge:conditions": [
+                            { "type": "neoforge:false" }
+                        ],
+                        "i": 1,
+                        "s": "test"
+                    }
+                    """);
+            // We expect a partial result with Optional.empty()
```
Comment (lukebemish) [inline tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java]: It now expects (and tests for) a normal result with Optional.empty()

**Review by lukebemish — APPROVED** [review]

File: tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java (hunk 4 above)
Comment (Technici4n) [reply tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java]: Thanks!

File: tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+import net.neoforged.fml.common.Mod;
+import net.neoforged.fml.event.lifecycle.FMLCommonSetupEvent;
+import net.neoforged.neoforge.common.conditions.ConditionalOps;
+import net.neoforged.neoforge.common.conditions.TrueCondition;
+import net.neoforged.neoforge.common.conditions.WithConditions;
+import org.apache.commons.lang3.mutable.MutableBoolean;
+import org.slf4j.Logger;
+
+@Mod("conditional_codec_test")
+public class ConditionalCodecTest {
```
Comment (sciwhiz12) [inline tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java:34]: Please provide a javadoc for this class which explains what feature this tests, how to test it, and what the expected results are from said test.

File: tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+    private static <T> void assertEquals(T expected, T actual) {
+        if (!Objects.equals(expected, actual)) {
+            throw new AssertionError("Expected " + expected + " but got " + actual);
+        }
+    }
+
+    private static void assertErrored(DataResult<?> result) {
+        MutableBoolean hasError = new MutableBoolean();
+        result.promotePartial(error -> hasError.setTrue());
+        assertEquals(true, hasError.booleanValue());
```
Comment (sciwhiz12) [inline tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java]: Why not check `DataResult.error().isPresent()`? That would avoid this use of `MutableBoolean` (and avoid the other work involved in `promotePartial`).

File: tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+
+    private static <T> String write(Codec<T> codec, T input) {
+        return GSON.toJson(codec.encodeStart(JsonOps.INSTANCE, input).get().left().get());
+    }
+
+    private static <T> void assertEquals(T expected, T actual) {
+        if (!Objects.equals(expected, actual)) {
+            throw new AssertionError("Expected " + expected + " but got " + actual);
+        }
+    }
```
Comment (sciwhiz12) [inline tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java:249]: We can replace this with `Assertions.assertEquals` from JUnit. The methods in `Assertions` throw a subclass of `AssertionError`, and have no dependencies on anything in the JUnit unit test APIs.

File: src/main/java/net/neoforged/neoforge/common/util/NeoForgeExtraCodecs.java [hunk 8]
```diff
@@ -107,22 +108,44 @@ public static <K, V> Codec<V> partialDispatchUnsafe(final Codec<K> keyCodec, fin
...
     }
 
-    public static <A> Decoder<List<A>> listDecoder(final Decoder<A> decoder) {
-        return Codec.of(Codec.unit(() -> {
-            throw new UnsupportedOperationException("Cannot encode with list decoder!");
-        }), decoder, decoder.toString()).listOf();
+    /**
+     * Creates a decoder invoking a callback for each element and the corresponding index in a list.
+     */
+    public static <A> Decoder<List<A>> listDecoderWithIndexConsumer(final Decoder<List<A>> decoder, ObjIntConsumer<A> consumer) {
```
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/common/util/NeoForgeExtraCodecs.java]: Personally, I'd name this something like `listDecoderWithIndexedPeek`, in reference to `Stream#peek` which looks at the elements of a stream without modifying them. However, this is okay as well.

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]

File: tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java (hunk 7 above)
Comment (Technici4n) [reply tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java:249]: Unfortunately this can't be done without NG support for adding entries to the legacy class path. JUnit gets yeeted by BootstrapLauncher.

File: tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java (hunk 6 above)
Comment (Technici4n) [reply tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java]: Not sure what I was thinking here...

File: src/main/java/net/neoforged/neoforge/common/util/NeoForgeExtraCodecs.java (hunk 8 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/common/util/NeoForgeExtraCodecs.java]: Done

File: tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java (hunk 7 above)
Comment (lukebemish) [reply tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java:249]: Assuming this is the same issue that currently makes using non-mod libraries a bit painful, I work around it with this: https://github.com/lukebemishprojects/BiomeSquisher/blob/ee39e65e25aed656f1dda7916b4af980eb039be3/neoforge/build.gradle#L57 - though no clue if that's something I'm supposed to be poking or not

Comment (Technici4n) [reply tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java:249]: I have it in AE2 too: https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/8854a324095b9df1a78906a579c3ac85f09c985f/build.gradle#L299, but that might not be a great fit for NeoForge itself... :P

Comment (sciwhiz12) [reply tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java:249]: If that is the case, there should be a comment there explaining the same.

Comment (Technici4n) [reply tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java:249]: Yeah, added a comment

**Comment by Technici4n** [comment]
@sciwhiz12 bump

**Review by sciwhiz12 — APPROVED** [review]
Approving but not merging, since someone needs to write an announcement about the change in the condition key as it's a breaking change and I lack the energy to do so.

Feel free to merge once an announcement's written up and ready to be posted. (Even if it's just a Discord announcement, for now.)
